DepositRegistrar.sol

```diff
@@ -0,0 +1,306 @@
+// Name registrar using deposits
+
+contract Deed {
+    // The Deed is a contract intended simply to hold ether
+    // It can be controlled only by the registrar and can only send ether back to the owner
+    address public owner;
+    address public bidder;
+    Registrar public registrar;
+    bytes32 public sealedBid;
+    bytes32 public hashedName;
+    uint public value;
+    address public burn = 0xdead;
+    uint public creationDate;
+    
+    modifier noEther {
+        if (msg.value > 0) throw;
+        _
+    } 
+    
+    modifier onlyRegistrar {
+        if (msg.sender != address(registrar)) throw;
+        _
+    }
+    
+    function Deed(address _bidder, bytes32 _sealedBid, address _registrar) {
+        bidder = _bidder;
+        registrar = Registrar(_registrar);
+        sealedBid = _sealedBid;
+        creationDate = now;
+    }
+    
+    function unsealBid(bytes32 _hashedName, address _owner, uint _value, bytes32 _salt) noEther onlyRegistrar {
+        if (sha3(_hashedName, _owner, _value, _salt) != sealedBid) throw;
+        hashedName = _hashedName;
+        owner = _owner;
+        value = _value;
+        if (!bidder.send(this.balance - _value)) throw;
+    }
+    
+    function closeDeed( uint burnAmount) noEther onlyRegistrar {
+        burn.send((burnAmount * this.balance)/1000);
+        selfdestruct(bidder);
+    }
+    
+    function setBalance(uint newValue) onlyRegistrar {
+        // Check if it has enough balance to set the value
+        if (this.balance < newValue) throw;
+        // Send the difference to the owner
+        if (!owner.send(this.balance - newValue)) throw;
+    }
+    
+    function transferOwnership(address newOwner) {
+        if (msg.sender != owner ) throw;
+        owner = newOwner;
+    }
+}
+
+contract Registrar {
+    mapping (bytes32 => entry) public hashedNames;
+    mapping (bytes32 => address) public sealedBids;
+    
+    enum Mode { Open, Auction, Owned }
+    uint auctionLength = 20 minutes;
+    uint revealPeriod = 5 minutes;
+    uint renewalPeriod = 2 hours;
+    uint M = 1000000; // just a multiplier to get more precision on averages
+    uint public averagePeriod;
+    uint public lastSinceNewName;
+    uint public minBid = 0.1 ether;    // TODO: figure out a better way to come up with this
```

Comment:
Problem with that approach is that miners can create transactions with a very low msg.gas, so someone could pay a miner to send a lot of 0 transactions and move the average way down